homeassistant/components/websocket_api/http.py

```diff
@@ -161,11 +163,31 @@ async def _writer(self) -> None:
                     messages.append(process if isinstance(process, str) else process())
                     messages_remaining -= 1
 
-                joined_messages = ",".join(messages)
-                coalesced_messages = f"[{joined_messages}]"
-                if debug_enabled:
-                    debug("%s: Sending %s", self.description, coalesced_messages)
-                await send_str(coalesced_messages)
+                coalesced_messages = f"[{','.join(messages)}]"
+                # We only check the length of the coalesced messages after
+                # combining them to avoid the overhead of checking the length
+                # of each message individually, and since the majority of
+                # the time we will never hit the max message size limit this
+                # is a good tradeoff.
+                if len(coalesced_messages) < max_message_size:
+                    if debug_enabled:
+                        debug("%s: Sending %s", self.description, coalesced_messages)
+                    await send_str(coalesced_messages)
+                    continue
+
+                logger.warning(
+                    "%s: Coalesced messages exceeded maximum size (%s/%s), this usually "
+                    "indicates an integration is sending too much data, one of the "
+                    "registries is too large, or the system cannot "
+                    "keep up; trying to send individually",
+                    self.description,
+                    len(coalesced_messages),
+                    max_message_size,
+                )
+                for message in messages:
+                    if debug_enabled:
+                        debug("%s: Sending %s", self.description, message)
+                    await send_str(message)
```

Comment:
Instead of either 1) sending all messages coalesced to one message or 2) sending messages one by one, should we coalesce messages until adding an additional message would exceed `_MAX_MESSAGE_SIZE`?

Reply:
I think its better to try to send them one at a time to give the system a bit of a chance to catch up rather than compressing a larger payload since and hope that the event loop stabilizes a bit more between awaits

Reply:
The thought process is sending the websocket messages is less important than whatever is going on with the system so the hope is that whatever is going on gets a chance to finish up and the websocket will get a chance to do its work after the system calms down. If we start sending large payloads, they have to be compressed which may make the system even more unstable.

Reply:
OK, but this PR is a bug fix intended to make sure we don't send too large messages.
Doing that in a more inefficient way than necessary, and hoping it somehow alleviates another problem, which is that the system is overloaded, doesn't seem right.

Reply:
If we think delivering the websockets messages is the highest priority, I'll refactor it to send the chunk messages into the largest possible allowed once it hits the limit

Reply:
I'm not saying delivering websocket messages is or isn't the highest priority, I'm saying coalescing websocket messages in an unnecessarily inefficient way does not seem like a reliable way to improve a problem where core is CPU starved.

Reply:
implemented that as 5beda4c530be36dec9e54d15a9223c01135ee51f